core/ipc/rpc/kafka/src/main/java/org/opennms/core/ipc/rpc/kafka/KafkaRpcServerManager.java

```diff
@@ -0,0 +1,280 @@
+    }
+
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    public void unbind(RpcModule module) throws Exception {
+        if (module != null) {
+            final RpcModule<RpcRequest, RpcResponse> rpcModule = (RpcModule<RpcRequest, RpcResponse>) module;
+            registerdModules.remove(rpcModule.getId());
+            stopConsumerForModule(rpcModule);
+        }
+    }
+
+    private void stopConsumerForModule(RpcModule<RpcRequest, RpcResponse> rpcModule) {
+        KafkaConsumerRunner kafkaConsumerRunner  = rpcModuleConsumers.remove(rpcModule);
+        LOG.info("stopped kafka consumer for module : {}", rpcModule.getId());
+        kafkaConsumerRunner.shutdown();
+    }
+
+    public void destroy() {
+
+    }
+
+    public void setConfigAdmin(ConfigurationAdmin configAdmin) {
+        this.configAdmin = configAdmin;
+    }
+
+    private class KafkaConsumerRunner implements Runnable {
+
+        private final KafkaConsumer<String, byte[]> consumer;
+        private final AtomicBoolean closed = new AtomicBoolean(false);
+        private String topic;
+        private RpcModule<RpcRequest, RpcResponse> module;
+
+        public KafkaConsumerRunner(RpcModule<RpcRequest, RpcResponse> rpcModule, KafkaConsumer<String, byte[]> consumer, String topic) {
+            this.consumer = consumer;
+            this.topic = topic;
+            this.module = rpcModule;
+        }
+
+        public void shutdown() {
+            closed.set(true);
+            consumer.wakeup();
+        }
+
+        @Override
+        public void run() {
+            try {
+                consumer.subscribe(Arrays.asList(topic));
+                LOG.info("subscribed to topic {}", topic);
+                while (!closed.get()) {
+                    ConsumerRecords<String, byte[]> records = consumer.poll(Long.MAX_VALUE);
+                    for (ConsumerRecord<String, byte[]> record : records) {  
+                        RpcMessageProtos.RpcMessage rpcMessage = RpcMessageProtos.RpcMessage.parseFrom(record.value());
+                        String rpcId = rpcMessage.getRpcId();
+                        boolean hasSystemId = rpcMessage.hasSystemId();
+                        String minionId = getMinionIdentity().getId(); 
+                        if (hasSystemId && !(minionId.equals(rpcMessage.getSystemId()))) {
+                            // directed RPC and not directed at this minion
+                            LOG.info("MinionIdentity {} doesn't match with systemId {}, ignore the request", minionId,
+                                    rpcMessage.getSystemId());
+                            continue;
+                        } 
+                        RpcRequest request = module.unmarshalRequest(rpcMessage.getRpcContent().toStringUtf8());
+                        if (hasSystemId) {
+                            // directed RPC, there may be more than one request with same request Id, cache and allow only one.
```

Comment:
What's the purpose of the value in the cache here? The RPC IDs should be unique, so provided there is an entry we shouldn't need to process it again.

We can also perform this logic prior to unmarshaling the request - this will save some cycles if we choose to ignore the request.